### DISCUSSION TITLE
Button glitching style

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#11188] [Q&A]
I'm creating a toggle button with a custom style. It's based over FluentAvalonia's styles.

It works fine in Light mode, but in Dark mode, when you hover in and out of the button, there's a glitch where it displays another color first before displaying the desired color, creating a blinking effect. Probably from FluentAvalonia's standard theme.

How can I solve this? Is redefining the whole ControlTheme the only way?

One issue I have with that is that, currently, all my app custom styles are defined in a Styles files, and ControlTheme would need to be set elsewhere in a Dictionary. I'd need to do project restructuring to do that.

```xaml
<Style Selector="Button.toolbar, ToggleButton.toolbar">
    <Setter Property="Foreground" Value="{DynamicResource ButtonForeground}" />
    <Setter Property="Padding" Value="2" />
    <Setter Property="BorderThickness" Value="0" />
    <Setter Property="Background" Value="Transparent" />
    <Setter Property="BorderBrush" Value="Transparent" />
    <Setter Property="HorizontalAlignment" Value="Center" />
    <Setter Property="VerticalAlignment" Value="Center" />
    <Setter Property="HorizontalContentAlignment" Value="Center" />
    <Setter Property="VerticalContentAlignment" Value="Center" />
    <Setter Property="Focusable" Value="False" />
    <Setter Property="Template">
        <ControlTemplate>
            <Border Background="{TemplateBinding Background}"
                    BorderBrush="{TemplateBinding BorderBrush}"
                    BorderThickness="{TemplateBinding BorderThickness}"
                    Padding="{TemplateBinding Padding}">
                <ContentPresenter Name="PART_ContentPresenter"
                                  Background="{TemplateBinding Background}"
                                  BorderBrush="{TemplateBinding BorderBrush}"
                                  BorderThickness="{TemplateBinding BorderThickness}"
                                  CornerRadius="{TemplateBinding CornerRadius}"
                                  ContentTemplate="{TemplateBinding ContentTemplate}"
                                  Content="{TemplateBinding Content}"
                                  Padding="{TemplateBinding Padding}"
                                  RecognizesAccessKey="True"
                                  TextBlock.Foreground="{TemplateBinding Foreground}"
                                  HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                                  VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}" />
            </Border>
        </ControlTemplate>
    </Setter>

    <Style Selector="^:checked /template/ ContentPresenter">
        <Setter Property="Background" Value="{DynamicResource ButtonBackground}" />
        <Setter Property="Foreground" Value="{DynamicResource ButtonForeground}" />
    </Style>
    <Style Selector="^:pointerover /template/ ContentPresenter#PART_ContentPresenter">
        <Setter Property="Background" Value="{DynamicResource ButtonBackgroundPointerOver}" />
        <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrushPointerOver}" />
        <Setter Property="Foreground" Value="{DynamicResource ButtonForegroundPointerOver}" />
    </Style>
    <Style Selector="^:pressed /template/ ContentPresenter#PART_ContentPresenter">
        <Setter Property="Background" Value="{DynamicResource ButtonBackgroundPressed}" />
        <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrushPressed}" />
        <Setter Property="Foreground" Value="{DynamicResource ButtonForegroundPressed}" />
    </Style>
    <Style Selector="^:disabled /template/ ContentPresenter#PART_ContentPresenter">
        <Setter Property="Background" Value="{DynamicResource ButtonBackgroundDisabled}" />
        <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrushDisabled}" />
        <Setter Property="Foreground" Value="{DynamicResource ButtonForegroundDisabled}" />
    </Style>
    
    <Style Selector="^ /template/ ContentPresenter">
        <Setter Property="FontFamily" Value="{TemplateBinding FontFamily}" />
    </Style>
    <Style Selector="^ /template/ ContentPresenter">
        <Setter Property="FontSize" Value="{TemplateBinding FontSize}" />
    </Style>
</Style>
```

**Comment by maxkatz6** [comment]
Can't repro in Sandbox with these styles.
https://user-images.githubusercontent.com/3163374/235336348-c93ebdf0-6734-4aa9-8a64-c93684ecd3d3.mp4


> How can I solve this? Is redefining the whole ControlTheme the only way?
> One issue I have with that is that, currently, all my app custom styles are defined in a Styles files, and ControlTheme would need to be set elsewhere in a Dictionary. I'd need to do project restructuring to do that.

I assume you also have copy of old fluent resources from avalonia repo.
It's not backward compatible with new theme variants and theme dictionaries.

**Reply by maxkatz6** [reply]
As an alternative, you can implement theme switching like it was working in 0.10 - remove one dictionary/style with old colors and add another dictionary/style with another colors. Basically, replacing resource dictionaries in runtime. It's less efficient and no support for scoping but works. And you can use `AvaloniaLocator.Current.GetService<IPlatformSettings>()` to get current OS colors and define default theme variant this way.

**Reply by maxkatz6** [reply]
Also, you don't really need to replace styles with control themes to get ThemeVariants work. But resources need to be updated.
Why did you copy them in the first place? If to customize colors, there is an easier option now.

**Reply by mysteryx93** [reply]
ah it's not a glitch with theme switching, Light/Dark modes work just fine

The visual glitch is when hovering over the control. Like background color being changed twice in a row.

**Reply by maxkatz6** [reply]
On the video I was hovering the button after the switch

**Reply by mysteryx93** [reply]
ah ok you're getting the "old" color, I'm replacing that color with another color. Will try do do a repro of the problem.

**Comment by mysteryx93** [comment]
Here's a repo of the togglebutton style glitch when hovering in dark mode

[Avalonia11-ToggleButtonGlitch.zip](https://github.com/AvaloniaUI/Avalonia/files/11362614/Avalonia11-ToggleButtonGlitch.zip)

**Reply by maxkatz6** [reply]
![image](https://user-images.githubusercontent.com/3163374/235413493-4cbdf3a4-c04b-4e2b-b6ec-8b796e948d81.png)

Not sure, but I see multiple pointer over styles. First one probably is from FluentAvalonia.
Though, it shouldn't cause these issues.

**Reply by maxkatz6** [reply]
Could be something with transitions as well.

**Reply by amwx** [reply]
Had a look at this - yes, buttons in Fluentv2 have a background transition on them. It appears there is an issue with animating a solid color brush with the Opacity set (not the alpha channel). I saw that `ButtonBackgroundPointerOver` was redefined as 
```xaml
    <SolidColorBrush x:Key="ButtonBackgroundPointerOver" Color="{DynamicResource SystemAccentColorLight2}"
                     Opacity=".30" />
```
so the pointerover animation is {Transparent (#00000000), Opacity=1} -> {SystemAccentColorLight2 (#FFxxxxxx), Opacity = 0.3}
That Opacity=0.3 is the issue. When I removed that, the animation interpolated correctly. It looks like the animator interpolates the color and Opacity components separately, I wonder if the correct way to do this is to multiply the alpha channel by the opacity and then animate the color only?
If you don't want the transition (or want a workaround), you can add another selector that sets the transitions property of the ContentPresenter to null.

**Reply by mysteryx93** [reply]
So is that a bug? I don't understand the issue with enough details to open a ticket.
